### PR TITLE
M3-4544: VLAN Panel in Linode Networking tab, Pt. 2

### PR DESCRIPTION
**Pull request by DevDW**
## Description
The first follow-up to https://github.com/linode/manager/pull/6950. This PR addresses the `readOnly` prop to disable the Remove VLAN button and formatting for the Linodes list. I'll keep this PR small and address the other items in a separate ticket.

## Type of Change
- Non breaking change ('update', 'change')

## Note to Reviewers
Coming in follow-up PRs:
- [ ]  "Attach a VLAN" button functionality
- [ ] Fix: when user clicks on a listed Linode, the VLAN panel should update to reflect the VLANs on that Linode

### CODE REVIEW

File: packages/manager/src/components/EntityTable/ListEntities_CMR.tsx [hunk 1]
```diff
@@ -59,6 +60,7 @@ export const ListEntities: React.FC<CombinedProps> = props => {
                       {paginatedAndOrderedData.map(thisEntity => (
                         <RowComponent
                           key={thisEntity.id}
+                          readOnly={readOnly}
```
Comment (DevDW) [inline packages/manager/src/components/EntityTable/ListEntities_CMR.tsx:63]: I did a brief run-through of other places in the app to see if this caused anything to break or generated console errors. What I checked seemed okay, but we will want to have all the CMR Row components that are connected to this accept and use the `readOnly` prop. That ties into M3-4575

File: packages/manager/src/features/linodes/LinodesDetail/LinodeNetworking/VLANPanel/VlanTableRow.tsx [hunk 2]
```diff
@@ -68,25 +54,26 @@ export const VlanTableRow: React.FC<CombinedProps> = props => {
...
             to={`/linodes/${linodeID}/networking`}
             data-testid="vlan-row-link"
           >
             {getLinodeLabel(linodeID)}
-            {(idx !== firstFour.length - 1 && data.length <= 4 && `, `) ||
-              (data.length > 4 && `, `)}
+            {(idx !== data.length - 1 &&
+              data.length <= MAX_LINODES_VLANATTACHED_DISPLAY &&
+              `, `) ||
+              (data.length > MAX_LINODES_VLANATTACHED_DISPLAY && `, `)}
```
Comment (DevDW) [inline packages/manager/src/features/linodes/LinodesDetail/LinodeNetworking/VLANPanel/VlanTableRow.tsx:71]: Had some difficulty removing the comma from being part of the link

File: packages/manager/src/components/EntityTable/EntityTable_CMR.tsx [hunk 3]
```diff
@@ -35,12 +35,22 @@ interface Props {
     order: OrderByProps['order'];
     orderBy: OrderByProps['orderBy'];
   };
+  readOnly?: boolean;
+  currentLinode?: number;
```
Comment (johnwcallahan) [inline packages/manager/src/components/EntityTable/EntityTable_CMR.tsx:39]: Remember, this is a generic component used for different types of entities. A prop called `currentLinode` doesn't belong here, since it's irrelevant in every case this component is used except for one (Linode VLANs).

One solution would be to add the LinodeID to the `vlanData` mapping in LinodeVlans.tsx (it's available on the LinodeInterface object).

I'd say the same thing for the `readOnly` property: better to let row components take care of it.

Related: the mapped vlanData in LinodeVlans.tsx data should get its own interface (`ExtendedVLAN` or something) since we've added fields to it that the row component doesn't know about.

**Review by johnwcallahan — CHANGES_REQUESTED** [review]

**Review by WilkinsKa1 — APPROVED** [review]
Looks good, nice work with the current Linode list handling logic!

File: packages/manager/src/features/linodes/LinodesDetail/LinodeNetworking/VLANPanel/VlanTableRow.test.tsx [hunk 4]
```diff
@@ -0,0 +1,23 @@
+import * as React from 'react';
+import { truncateAndJoinJSXList } from './VlanTableRow';
+import { Link } from 'react-router-dom';
+
+describe('truncateAndJoinJSXList utility function', () => {
```
Comment (johnwcallahan) [inline packages/manager/src/features/linodes/LinodesDetail/LinodeNetworking/VLANPanel/VlanTableRow.test.tsx:5]: This function doesn't quite work in all cases, including the test case here. As an example, I have four Linodes attached to this VLAN, and I changed the max to 2:
Screen Shot 2020-10-08 at 2 24 10 PM

Examining the result of L15, you can see that something similar happens with the test case here. You definitely want to check the corner cases with a function like this: `length == max-1`, `length == max`, `length == max+1`, etc.

File: packages/manager/src/features/linodes/LinodesDetail/LinodeNetworking/VLANPanel/VlanTableRow.test.tsx [hunk 5]
```diff
@@ -0,0 +1,23 @@
...
+  const list = [0, 1, 2, 3];
+  const jsxList = list.map(item => {
+    return (
+      <Link key={item} to={`/test`}>
```
Comment (johnwcallahan) [inline packages/manager/src/features/linodes/LinodesDetail/LinodeNetworking/VLANPanel/VlanTableRow.test.tsx:9]: No need to use a `<Link />` which complicates the test, you can just do a one-level span, something like `<span key={item}>{item}</span>;`

File: packages/manager/src/features/linodes/LinodesDetail/LinodeNetworking/VLANPanel/VlanTableRow.test.tsx [hunk 6]
```diff
@@ -0,0 +1,23 @@
...
+        {item}
+      </Link>
+    );
+  });
+
+  const truncatedList = truncateAndJoinJSXList(jsxList, 3);
+  const truncatedText = [' ', 'plus ', 1, ' more'];
+
+  it('should have truncated text that reads " plus 1 more"', () => {
+    expect(truncatedList.props.children[1].props['children']).toEqual(
```
Comment (johnwcallahan) [inline packages/manager/src/features/linodes/LinodesDetail/LinodeNetworking/VLANPanel/VlanTableRow.test.tsx:19]: Nice job figuring out how to introspect these React elements;.

File: packages/manager/src/features/linodes/LinodesDetail/LinodeNetworking/VLANPanel/VlanTableRow.tsx [hunk 7]
```diff
@@ -3,33 +3,20 @@ import { APIError } from '@linode/api-v4/lib/types';
...
-    alignItems: 'center',
-    whiteSpace: 'nowrap'
-  }
-}));
-
-export type CombinedProps = VLAN & ActionHandlers;
+export const MAX_LINODES_VLANATTACHED_DISPLAY = 50;
+
+interface Props {
+  readOnly: boolean;
```
Comment (johnwcallahan) [inline packages/manager/src/features/linodes/LinodesDetail/LinodeNetworking/VLANPanel/VlanTableRow.tsx:16]: This prop is actually coming from the `VlanData`, right? L19 should be `Props & VlanData & ActionHandlers` I think. You'll have circular imports though, so you need to declare the VlanData interface somewhere else.

File: packages/manager/src/features/linodes/LinodesDetail/LinodeNetworking/VLANPanel/LinodeVLANs.tsx [hunk 8]
```diff
@@ -271,3 +274,10 @@ export const getInterfaceName = (
   }
   return null;
 };
+
+interface VlanData extends VLAN {
+  ip_address: string;
+  interfaceName: string | null;
+  currentLinode: number;
```
Comment (johnwcallahan) [inline packages/manager/src/features/linodes/LinodesDetail/LinodeNetworking/VLANPanel/LinodeVLANs.tsx:281]: I'd recommend calling this `currentLinodeID`... this name implies it's of type `Linode`.

File: packages/manager/src/features/linodes/LinodesDetail/LinodeNetworking/VLANPanel/VlanTableRow.tsx [hunk 9]
```diff
@@ -119,13 +109,43 @@ export const VlanTableRow: React.FC<CombinedProps> = props => {
...
+  add a comma after the JSX element.
+  2. If the length of the list is greater than the max, add truncation text at the end.
+  */
+
+  return (
+    <>
+      {JSXList.map((item, idx) => (
+        <span key={idx}>
+          {item}
+          {(idx !== count - 1 && count <= max && `, `) || (count > max && `, `)}
```
Comment (johnwcallahan) [inline packages/manager/src/features/linodes/LinodesDetail/LinodeNetworking/VLANPanel/VlanTableRow.tsx:137]: I'd strongly recommend against this type of one-liner here; multi-line, more verbose logic is more readable in a situation like this. One thing that would make it cleaner is if you sliced the list ahead of time.

**Review by johnwcallahan — CHANGES_REQUESTED** [review]

File: packages/manager/src/features/linodes/LinodesDetail/LinodeNetworking/VLANPanel/VlanTableRow.test.tsx [hunk 10]
```diff
@@ -1,23 +1,34 @@
...
+      longJSXList,
+      max
+    );
+    expect(noTruncatedTextListWhenEqual.props.children.length).toEqual(max);
+  });
 
-  it('should have truncated text that reads " plus 1 more"', () => {
-    expect(truncatedList.props.children[1].props['children']).toEqual(
-      expect.arrayContaining(truncatedText)
+  it('should have truncated text that reads " plus [X] more" if the max is 5', () => {
```
Comment (johnwcallahan) [inline packages/manager/src/features/linodes/LinodesDetail/LinodeNetworking/VLANPanel/VlanTableRow.test.tsx:32]: More precisely, "... if the max is less than the length of the list".

File: packages/manager/src/features/linodes/LinodesDetail/LinodeNetworking/VLANPanel/VlanTableRow.tsx [hunk 11]
```diff
@@ -122,30 +118,47 @@ export const truncateAndJoinJSXList = (
...
         <span data-testid="truncated-text">
-          {` `}plus {count - max} more
+          {` `}plus {remainingItemCount} more
         </span>
-      )}
-    </>
-  );
+      </>
+    );
+  }
```
Comment (johnwcallahan) [inline packages/manager/src/features/linodes/LinodesDetail/LinodeNetworking/VLANPanel/VlanTableRow.tsx]: Great job with this. IMO it's worth the extra lines of code and a little duplication... so much more readable.

**Review by johnwcallahan — APPROVED** [review]
LGTM, assuming the splice issue will be addressed in the next PR. Nice job.
